### PR TITLE
Tests/notebooks

### PR DESCRIPTION
**Pull request by akucukelbir**
We'd like to test (most of) our shiny new notebooks.

This addresses: https://github.com/blei-lab/edward/issues/224

I'm basically replicating the great solution of: https://github.com/GPflow/GPflow/pull/374

TODO
- [x] make sure it passes on Python 2.7
- [x] debug any travis issues (if any)

### CODE REVIEW

File: tests/test-notebooks/test_notebooks.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                             kernel_name=pythonkernel,
+                             interrupt_on_timeout=True)
+    lfiles = glob.glob(nbpath + "*.ipynb")
+    for notebook_filename in lfiles:
+      if(os.path.basename(notebook_filename) not in blacklist):
+        t = time.time()
+        self._execNotebook(ep, notebook_filename, nbpath)
+        print(notebook_filename, 'took %g seconds.' % (time.time() - t))
+
+
```
Comment (dustinvtran) [inline tests/test-notebooks/test_notebooks.py]: all other unit tests have only one blankline here

File: tests/test-notebooks/test_notebooks.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import glob
+import traceback
+import unittest
+import sys
+import time
+import os
+
+
+class TestNotebooks(unittest.TestCase):
+  def _execNotebook(self, ep, notebook_filename, nbpath):
```
Comment (dustinvtran) [inline tests/test-notebooks/test_notebooks.py]: all other unit tests have a blank line between the first class method and the `class` line. also we use snake case, `test_notebooks` and `_exec_notebook`.

File: tests/test-notebooks/test_notebooks.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      except CellExecutionError:
+        print('-' * 60)
+        traceback.print_exc(file=sys.stdout)
+        print('-' * 60)
+        self.assertTrue(False,
+                        'Error executing the notebook %s. See above for error.'
+                        % notebook_filename)
+
+  def test_all_notebooks(self):
+    ''' Test all notebooks except blacklist. '''
```
Comment (dustinvtran) [inline tests/test-notebooks/test_notebooks.py]: we use `"""` for docstrings

File: tests/test-notebooks/test_notebooks.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import nbformat
+from nbconvert.preprocessors import ExecutePreprocessor
+from nbconvert.preprocessors.execute import CellExecutionError
+
+import glob
+import traceback
+import unittest
+import sys
+import time
+import os
```
Comment (dustinvtran) [inline tests/test-notebooks/test_notebooks.py]: can you alphabetize the import statements, then add a blank line, then add the `from` import statements? this mimicks all other python files

File: tests/test-notebooks/test_notebooks.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class TestNotebooks(unittest.TestCase):
```
Comment (dustinvtran) [inline tests/test-notebooks/test_notebooks.py]: following the other unit tests, should we use `tf.test.TestCase` instead?

File: tests/test-notebooks/test_notebooks.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        print('-' * 60)
+        traceback.print_exc(file=sys.stdout)
+        print('-' * 60)
+        self.assertTrue(False,
+                        'Error executing the notebook %s. See above for error.'
+                        % notebook_filename)
+
+  def test_all_notebooks(self):
+    ''' Test all notebooks except blacklist. '''
+    blacklist = ['gan.ipynb', 'iclr2017.ipynb']
```
Comment (dustinvtran) [inline tests/test-notebooks/test_notebooks.py]: We can remove `iclr2017.ipynb` off the black list. It's not expensive and should run to verify all the syntax works.

File: tests/test-notebooks/test_notebooks.py (hunk 4 above)
Comment (akucukelbir) [inline tests/test-notebooks/test_notebooks.py]: sure thing.

File: tests/test-notebooks/test_notebooks.py (hunk 5 above)
Comment (akucukelbir) [inline tests/test-notebooks/test_notebooks.py]: i thought about this. we don't depend on any specific tensorflow functionality here (like having access to a session). this is why i kept it as a `unittest`

File: tests/test-notebooks/test_notebooks.py (hunk 2 above)
Comment (akucukelbir) [inline tests/test-notebooks/test_notebooks.py]: great catch

File: tests/test-notebooks/test_notebooks.py (hunk 3 above)
Comment (akucukelbir) [inline tests/test-notebooks/test_notebooks.py]: done

File: tests/test-notebooks/test_notebooks.py (hunk 6 above)
Comment (akucukelbir) [inline tests/test-notebooks/test_notebooks.py]: okay. i wasn't sure if this was a long-term notebook. if so, we may want to rename it to something a little bit more descriptive.

File: tests/test-notebooks/test_notebooks.py (hunk 1 above)
Comment (akucukelbir) [inline tests/test-notebooks/test_notebooks.py]: done

**Comment by akucukelbir** [comment]
re: questions:

1. i was under the impression that we would be removing examples that are now represented in a notebook. if so, i think it would make more sense to keep a single copy of the data and to keep it near the notebooks. 
2. just so that i'm on the same page: are you asking whether we can do a pep8 check on the code cells in our notebooks?

**Comment by dustinvtran** [comment]
1. It would be nice to have the data closer to the notebooks. At the same time, it would also be nice to have only one directory housing any external data.
2. Correct.

File: tests/test-notebooks/test_notebooks.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        print('-' * 60)
+        self.assertTrue(False,
+                        'Error executing the notebook %s. See above for error.'
+                        % notebook_filename)
+
+  def test_all_notebooks(self):
+    """ Test all notebooks except blacklist. """
+    blacklist = ['gan.ipynb']
+    pythonkernel = 'python' + str(sys.version_info[0])
+    nbpath = '../../docs/notebooks/'
```
Comment (dustinvtran) [inline tests/test-notebooks/test_notebooks.py]: Do you know how this nbpath works? My impression was that travis calls pytest within the repo's director and not within `tests/test-notebooks/`, which is what assume this nbpath is referring to.

**Comment by akucukelbir** [comment]
1. how do you suggest we restructure the directories?

i'm not sure why the notebooks are within `\docs\`, which to me only involves things relating to our website.

perhaps we want to port `docs\notebooks` into the `examples` directory and have the `data` folder there? 

perhaps we should rename `\docs` into `\website`?

2. i'm looking into it. might be more of a headache than what it's worth though, especially when we're writing/developing new notebooks. but let me be get back to you on this.

**Comment by dustinvtran** [comment]
> i'm not sure why the notebooks are within` \docs\`, which to me only involves things relating to our website.

It's only in there for legacy reasons. I agree we should move it. 

I'd prefer to separate the `notebooks/` directory from the `examples/` directory. If we make `notebooks/` a top-level directory, then I'd be okay with having two `data/` subdirectories, one for `examples/` and one for `notebooks/`.

2. I think it's important to lay out good practices regardless of format. That said, it's not necessary for this PR though—we can look into it later as part of another PR.

**Comment by akucukelbir** [comment]
1. ok let's take care of that in a separate PR.

2. this makes more sense to take care of here. are you open to running another command beyond `pytest` during testing? i think i can do this fairly easily with `flake8` for example

**Comment by akucukelbir** [comment]
Ah I missed that. Sure. `pep8` also works. The main question is whether we're okay with call(s) to `pep8` in travis. (Right now we only call `pytest`)

**Comment by dustinvtran** [comment]
It's worth thinking about how to simplify but not a priority. (For example, do we know how the pytest command currently checks PEP8 in all files? Is it possible to pipe the nbconversion into it as well?)

**Comment by akucukelbir** [comment]
I looked into it briefly and it looked kinda complicated, mainly due to how `pytest` looks for files to test.

**Comment by dustinvtran** [comment]
Bump. Let me know if I can help in any way.

**Pull request merged**